### PR TITLE
Create USACO jan20-race.mdx

### PR DESCRIPTION
**Pull request by jessechoe10**

### CODE REVIEW

File: solutions/USACO jan20-race.mdx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		if(k-curArea == 0){
+			cout << 2*y - x << endl;
+		} else if(k-curArea <=y ){
+			cout << 2*y - x+1 << endl;
+		} else {
+			cout << 2*y - x+2 << endl;
+		}
+	}
+
+}
```
Comment (Farrius) [inline solutions/USACO jan20-race.mdx:114]: Seems good! But there is already a nice explanation in the official editorial and it use and c++, so I think there is no need to add another solution. Nevertheless wait for other reviews.

File: solutions/USACO open18-multiplayermoo.mdx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    cout << ans1 << endl;
+    FOR(i, 1, n){
+        FOR(j, 1, n){
+            cnt2 = 0;
+            if(ans2 == n*n) break;
+            if(j > 0 && board[i][j] != board[i][j-1]){
+                FF2(i, j, board[i][j], board[i][j-1]);
+                ans2 = max(ans2, cnt2);
+                cnt2 = 0;
```
Comment (Farrius) [inline solutions/USACO open18-multiplayermoo.mdx:123]: I think is better to put cnt2 before FF2, just like the other FF. And then you can remove the previous cnt2 = 0.

File: solutions/USACO open18-multiplayermoo.mdx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            cnt1 = 0;
+            FF(i, j, board[i][j]);
+            ans1 = max(cnt1, ans1);
+        }
+    }
+    cout << ans1 << endl;
+    FOR(i, 1, n){
+        FOR(j, 1, n){
+            cnt2 = 0;
+            if(ans2 == n*n) break;
```
Comment (Farrius) [inline solutions/USACO open18-multiplayermoo.mdx:119]: There is no need to put this if statement.

File: solutions/USACO open18-multiplayermoo.mdx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                FF2(i, j, board[i][j], board[i][j-1]);
+                ans2 = max(ans2, cnt2);
+                cnt2 = 0;
+                memset(visited2, false, sizeof(visited2));
+            }
+        }
+    }
+    cout << ans2 << endl;
+}
+```
```
Comment (Farrius) [inline solutions/USACO open18-multiplayermoo.mdx:130]: I found the explanation really good and the code too!

File: solutions/USACO open18-multiplayermoo.mdx (hunk 3 above)
Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:119]: If I remove this if statement, then I TLE on one of the TCS.

File: solutions/USACO open18-multiplayermoo.mdx (hunk 4 above)
Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:130]: Thank you!

**Review by darren-yao — CHANGES_REQUESTED** [review]
@jessechoe10 Please address all of Farrius's comments, then ping me on Discord when you're done.

File: solutions/USACO open18-multiplayermoo.mdx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+This problem is an example of a "floodfill" problem. This problem is best solved when divided into two subproblems.
+
+The first subproblem is to find the maximum number of adjacent cells comprising of the same ID number. To calculate the maximum number of adjacent cells, we can iterate over each cell in the grid and count the maximum number of "connected" cells via a recursive floodfill approach. To avoid recomputation, we can store a boolean array of the cells we "visited" and only run the floodfill on this cell if and only if that cell hasn't been visited.
+
+The second subproblem is to find the maximum number of adjacent cells comprising of exactly two distinct ID numbers (note: if two different ID numbers are adjacent to each other, then they are considered "connected"). To calculate the maximum number of "connected" cells, we can iterate over each cell in the grid and compute this by choosing any two adjacent cells as the ID's. From here, we can run a floodfill and find the maximum number of "connected" cells over all possible cells. As used in the previous subproblem, we can use a boolean array to prevent recomputation.
+
+**Time Complexity:** $\mathcal{O}(N^2)$
```
Comment (bqi343) [inline solutions/USACO open18-multiplayermoo.mdx:18]: if this solution was actually N^2 then you wouldn't be having issues with the TL. This is N^4 ...

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan20-race.mdx (hunk 1 above)
Comment (bqi343) [reply solutions/USACO jan20-race.mdx:114]: well, this is O(1) per query

File: solutions/USACO jan20-race.mdx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		if(x >= p){
+			if(reached){
+				cout << p << endl;
+			} else {
+				cout << p+1 << endl;
+			}
+			continue;
+		}
+		
+		int y = sqrt(double(((2*k+pow(x, 2)-x) / 2)));
```
Comment (bqi343) [inline solutions/USACO jan20-race.mdx:101]: don't need `double` on this line

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan20-race.mdx (hunk 6 above)
Comment (jessechoe10) [reply solutions/USACO jan20-race.mdx:101]: Oops, it seems like I put the wrong code after looking at it. MB. I have another version using simulation.

File: solutions/USACO open18-multiplayermoo.mdx (hunk 5 above)
Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: Isn't it N^2 bc of the visited array?

Comment (bqi343) [reply solutions/USACO open18-multiplayermoo.mdx:18]: but you reset the visited array

Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: is it cuz of the memset that makes it N^4? if it is mb

File: content/2_Bronze/Intro_Greedy.mdx [hunk 7]
```diff
@@ -90,7 +90,8 @@ export const problems = {
 			'Very Hard',
 			true,
 			['Greedy'],
-			'Greedily increment/decrement Bessies speed to fit the conditions until her total distance exceeds K.'
+			'Greedily increment/decrement Bessies speed to fit the conditions until her total distance exceeds K.',
+			'usaco-989',
```
Comment (frodakcin) [inline content/2_Bronze/Intro_Greedy.mdx:94]: Not linking properly.

Comment (frodakcin) [reply content/2_Bronze/Intro_Greedy.mdx:94]: (link: https://usaco-guide-9hutey16s.vercel.app/bronze/intro-greedy?lang=cpp)

File: solutions/USACO open18-multiplayermoo.mdx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            ans1 = max(cnt1, ans1);
+        }
+    }
+    cout << ans1 << endl;
+    FOR(i, 1, n){
+        FOR(j, 1, n){
+            cnt2 = 0;
+            if(ans2 == n*n) break;
+            if(j > 0 && board[i][j] != board[i][j-1]){
+	    	cnt2 = 0;
```
Comment (frodakcin) [inline solutions/USACO open18-multiplayermoo.mdx:121]:
```suggestion
                cnt2 = 0;
```

seems to be tab vs space

**Review by frodakcin — CHANGES_REQUESTED** [review]

File: content/2_Bronze/Intro_Greedy.mdx (hunk 7 above)
Comment (jessechoe10) [reply content/2_Bronze/Intro_Greedy.mdx:94]: Can you remove this solution altogether? I'll make another PR with my simulation code a bit later. :D

File: solutions/USACO open18-multiplayermoo.mdx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+This problem is an example of a "floodfill" problem. This problem is best solved when divided into two subproblems.
+
+The first subproblem is to find the maximum number of adjacent cells comprising of the same ID number. To calculate the maximum number of adjacent cells, we can iterate over each cell in the grid and count the maximum number of "connected" cells via a recursive floodfill approach. To avoid recomputation, we can store a boolean array of the cells we "visited" and only run the floodfill on this cell if and only if that cell hasn't been visited.
+
+The second subproblem is to find the maximum number of adjacent cells comprising of exactly two distinct ID numbers (note: if two different ID numbers are adjacent to each other, then they are considered "connected"). To calculate the maximum number of "connected" cells, we can iterate over each cell in the grid and compute this by choosing any two adjacent cells as the ID's. From here, we can run a floodfill and find the maximum number of "connected" cells over all possible cells. As used in the previous subproblem, we can use a boolean array to prevent recomputation.
+
+**Time Complexity:** $\mathcal{O}(N^4)$
```
Comment (bqi343) [inline solutions/USACO open18-multiplayermoo.mdx:18]: this solution isn't supposed to pass though ... (it's pretty easy to break)

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan20-race.mdx (hunk 1 above)
Comment (bqi343) [reply solutions/USACO jan20-race.mdx:56]: it's no longer O(1) per query, I agree with @Farrius that there's no need to add this

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/2_Bronze/Intro_Greedy.mdx (hunk 7 above)
Comment (frodakcin) [reply content/2_Bronze/Intro_Greedy.mdx:93]: Do you want to revert `content/2_Bronze/Intro_Greedy.mdx` and `solutions/USACO jan20-race.mdx`?

File: solutions/USACO open18-multiplayermoo.mdx (hunk 9 above)
Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: Well idk, it passes all the USACO tcs and the official solution is way too ugly.

Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: > this solution isn't supposed to pass though ... (it's pretty easy to break)

Does this mean that every singe test case on USACO was bad?

File: content/2_Bronze/Intro_Greedy.mdx (hunk 7 above)
Comment (jessechoe10) [reply content/2_Bronze/Intro_Greedy.mdx:93]: Yeah.

File: solutions/USACO open18-multiplayermoo.mdx (hunk 9 above)
Comment (bqi343) [reply solutions/USACO open18-multiplayermoo.mdx:18]: yes ...

Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: So we can't put this sol despite passing 10/10 tcs?

Comment (bqi343) [reply solutions/USACO open18-multiplayermoo.mdx:18]: yes

Comment (jessechoe10) [reply solutions/USACO open18-multiplayermoo.mdx:18]: :(

**Comment by bqi343** [comment]
closing this, described solution only passes because the data was generated quite poorly. feel free to reopen if you have smth with better time complexity ...

**Pull request closed without merging**